include/deal.II/base/smartpointer.h

```diff
@@ -366,11 +369,14 @@ SmartPointer<T, P>::operator=(const SmartPointer<T, Q> &tt)
   if (&tt == this)
     return *this;
 
+  // Let us unsubscribe from the current object
   if (pointed_to_object_is_alive && t != nullptr)
     t->unsubscribe(&pointed_to_object_is_alive, id);
-  t = static_cast<T *>(tt);
+
+  // Then reset to the new object, and subscribe to it
+  t = (tt != nullptr ? tt.get() : nullptr);
```

Comment:
Does the conversion operator
```cpp
t = tt;
```
work as we would expect here? I'd prefer that over the ternary.

Reply:
I found that hard to read because while they look the same from their names, `t` is a pointer and `tt` is a `SmartPointer`. When you say `t=tt`, you use the implicit conversion on the right hand side, and I think that makes it harder to read. I'll change it back if you want me to, but I don't think it is easier to read.

Reply:
I don't feel that strongly about it so lets go with the present version.